### DISCUSSION TITLE
Convert ListConfig objects with no structure to list type within DictConfig?

### DISCUSSION THREAD

**Discussion by LarsHill** [omry/omegaconf#848] [Q&A]
Hi,

Using omegaconf as a wrapper around config dicts led to a few problems with respect to the ListConfig type. Some well known external libraries, i.e. huggingface's transformers check specifically for the inbuild list type for certain function inputs. 

For example, when adding additional special tokens to a Tokenizer object, these special tokens have to be of type list. Since I register these in a config, the resulting type within the parsed DictConfig object is ListConfig (does not inherit from list) and the transformer library raises an error. 

I saw a related issue wrt PyTorch Lighting and the answer was that it is not recommended to inherit from primitive containers and that it is the responsibility of external libraries to check for e.g. the MutualSequence type instead of the inbuild list type. Unfortunately many libraries don't do that. 

An other problem I encountered was the following:
In my config I register a list of hidden layer dimensions for a simple MLP model, e.g. 
```yaml
hidden_dims: [512, 256]
```
In my model I have to add the input and output dimensions (data and task dependent) to this object. However a simple
`hidden_dims = [128] + hidden_dims + [2]`
does not work since `hidden_dims` is of type ListConfig and it cannot be added to a list type object.

```python
from omegaconf import DictConfig
d = {'a': 1, 'b': [1,2,3]}
d_cfg = DictConfig(d)
hidden_dims = [10] + d_cfg.b + [1]
=> TypeError: can only concatenate list (not "ListConfig") to list
```

Since I don't want to manually convert normal list type hyper params to the inbuild list types every time some error is raised, I look for a way to convert all simple ListConfig objects that don't contain a DictConfig structure to the normal list type.

Hence, I tried this simple test, but got:

```Python
from omegaconf import DictConfig

d = {'a': 1, 'b': [1, 2, 3]}
d_cfg = DictConfig(d)
print(type(d_cfg.b))
=> ListConfig

d_cfg.b = list(d_cfg.b)
print(type(d_cfg.b))
=> ListConfig
```

In the second print statement I expected to see type list, but apperantly the list object is wrapped dynamically to a ListConfig object when it is newly assigned.

So is there any other way of avoiding the ListConfig type for e.g. lists that don't have any further nesting?

I think an option like this would be great because it would allow the user to use all the benefits of DictConfig but he can expect that his list hyperparameters don't cause any problems in external libraries or with operations that work with the list type but not with the ListConfig type. 

Thanks for your help and answers :)

**Comment by Jasha10** [comment]
The OmegaConf container types (ListConfig and DictConfig) quack like lists and dicts, respectively. However, as you pointed out, many libraries rely on python-native lists and dicts, and do not play as well with MutableSequences and MutableMappings.

For these situations, I'd recommend calling [`OmegaConf.to_object`](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-to-object) to recursively convert your OmegaConf object to a plain python object.
```python
>>> d = {'a': 1, 'b': [1, 2, 3]}
>>> d_cfg = DictConfig(d)
>>> d_obj = OmegaConf.to_object(d_cfg)
>>> print(type(d_obj))
<class 'dict'>
>>> print(type(d_obj["b"]))
<class 'list'>
```

---

> In my model I have to add the input and output dimensions (data and task dependent) to this object. However a simple `hidden_dims = [128] + hidden_dims + [2]` does not work since `hidden_dims` is of type ListConfig and it cannot be added to a list type object.

I have just opened up a feature request (#849) to address this.
Once implemented, the sum `[128] + hidden_dims + [2]` will result in a ListConfig (not a list).

---

> In the second print statement I expected to see type list, but apperantly the list object is wrapped dynamically to a ListConfig object when it is newly assigned.
> So is there any other way of avoiding the ListConfig type for e.g. lists that don't have any further nesting?

It's not possible to do this with OmegaConf currently. If you want a list (not a ListConfig), just use `OmegaConf.to_object`.

The motivation for converting lists to ListConfigs is so that extra "metadata" can be tracked (beyond what a plain list would be able to do). Examples of this metadata include typing information (enabling runtime type-checking for [structured configs](https://omegaconf.readthedocs.io/en/2.1_branch/structured_config.html)) and keeping track of the parent container (which is necessary so that [interpolations to a parent node](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#config-node-interpolation) will work).

In practice, I find that OmegaConf is most useful during the "configuration" phase; often, practitioners end up converting "back to plain python" when it's time to run the business logic.

**Reply by LarsHill** [reply]
Thanks for the reply. I am well aware of the option to simply call `OmegaConf.to_object()`. 

But calling it on a DictConfig object removes the advantage of nice attribute access of nested config arguments.
And this is at the moment the main reason why I use omegaconf. Hence, there would be no need to use it at all, given one does not use the other configuration features (metadata, etc.) it offers.

Calling `OmegaConf.to_object()` on a ListConfig type when needed is also not a nice solution. Because as a user I would expect, that my hyper parameter list specified in the config fully behaves like a normal list. And at the moment this is not the case. 
So I figured it would be nice to have an option to indicate to OmegaConf which lists in the config need additional metadata and hence use the ListConfig type and which lists are simple leave nodes that can be treated as standard lists when parsing.

Of course if ListConfig would inherit from list and has all its capabilities, it would behave as expected in all cases and there would be no problem :)

**Reply by Jasha10** [reply]
> And this is at the moment the main reason why I use omegaconf.

I see.

> Calling `OmegaConf.to_object()` on a ListConfig type when needed is also not a nice solution. Because as a user I would expect, that my hyper parameter list specified in the config fully behaves like a normal list. And at the moment this is not the case.

The good news here is that the patch for `[128] + hidden_dims + [2]` is very simple. I've opened a PR (#850) and expect I'll be able to merge it soon.

> So I figured it would be nice to have an option to indicate to OmegaConf which lists in the config need additional metadata and hence use the ListConfig type and which lists are simple leave nodes that can be treated as standard lists when parsing.
> 
> Of course if ListConfig would inherit from list and has all its capabilities, it would behave as expected in all cases and there would be no problem :)

Duck-typed compatibility with lists is a goal wherever possible! 
That being said, the conversion from list to ListConfig is heavily baked into OmegaConf's implementation; disabling this conversion would be a big project...

**Reply by Jasha10** [reply]
#850 is now merged.

**Reply by LarsHill** [reply]
Thanks for the clarification. I can imagine that adding such functionality would be a major project. 

In the meantime I encountered an other situation where DictConfig leads to an error and does not behave like a regular dict:
Checking whether dict A is a subset of dict B works if both objects are of type dict and it also works if both objects are of type DictConfig.

However, it does not work if one object is of type DictConfig and the other is of type dict. See this minimal example:

```python
from omegaconf import OmegaConf

d1 = {'a': 1, 'b': [1, 2, 3]}
d2 = {'a': 1, 'b': [1, 2, 3]}
d1_cfg = OmegaConf.create(d1)
d2_cfg = OmegaConf.create(d2)

if d1.items() <= d2.items():
    print(True)

if d1_cfg.items() <= d2_cfg.items():
    print(True)

if d1.items() <= d2_cfg.items():
    print(True)
```

This Code prints
```
True
True
TypeError: '<=' not supported between instances of 'dict_items' and 'list'
```

Is there a reason for `d1_cfg.items()` being of type list instead of dict_items? For compatibility reasons with dicts the type "dict_items" would be preferable, no?

**Reply by Jasha10** [reply]
I see.
For `d1_cfg.items()`, returning a `list` instead of `dict_items` makes for a simpler implementation.
I'm sorry to say that I don't think there's much we can do about this incompatibility.
Although the dictionary views (i.e. the return types from `dict.{keys,values,items}` [are documented](https://docs.python.org/3.3/library/stdtypes.html#dict-views), there is no public python API (that I know of) for creation of such views (without calling the actual method on a `dict` instance).
Moreover, if we were to return `dict_items` from `DictConfig.items`, this could be a breaking change for existing user code that assumes `DictConfig.items` returns a list.

A possible workaround for compatibility between `dict.items` and `DictConfig.items` would be to convert the `dict_items` instance to a list before comparing:
```python
if list(d1.items()) <= d2_cfg.items():
    print(True)
```

**Reply by Jasha10** [reply]
I think better compatibility between dictionary views and lists (for inequality comparison) is something that could be implemented on the CPython side.
I'd create a feature request on the [CPython bug tracker](https://bugs.python.org), but I'm having trouble logging in there.

**Reply by LarsHill** [reply]
The `MutableMapping` interface inherits from `Mapping` and thus an `items()` method is implemented already when inheriting from `MutableMapping`. 
You overwrite the default `items()` in OmegaConf, probably due to some metadata handling. Is it really necessary though? 

Nevertheless, when checking the `MutableMapping` implementation, it turns out that `items()` returns an `ItemsView` object which is defined here: https://github.com/python/cpython/blob/d7c68639795a576ff58b6479c8bb34c113df3618/Lib/_collections_abc.py#L884

Maybe something similar can be returned in the overwritten `items()`?

Regardless of backward compatibility with previous omegaconf versions I still think it should be the goal of `DictConfig` to behave as close to a dict as possible.

**Reply by Jasha10** [reply]
>  You overwrite the default items() in OmegaConf, probably due to some metadata handling. Is it really necessary though?

It might be possible to use MutableMapping's `items` method, but it would require major refactoring. It is something that probably would not happen for OmegaConf v2 due to backwards compatibility issues.

**Reply by Jasha10** [reply]
OmegaConf enables many use-cases (e.g. [Hydra](https://hydra.cc), which uses OmegaConf as a backend) besides just the `__getattr__` lookup for attributes.

If you need perfect compatibility with `dict`, it will probably be easier for you to define your own subclass of `MutableMapping`.

**Reply by LarsHill** [reply]
In case others are reading this discussion:

I indeed created my own subclass which behaves exactly as a `dict` with the addition of enabling (nested) attribute-style key access/assignment (dot notation) and autocompletion. 

I named it [metadict](https://github.com/LarsHill/metadict) and it is available via `pip`.

**Reply by Jasha10** [reply]
Cool!

**Reply by Jasha10** [reply]
FYI in PR #863 we changed the `DictConfig.keys` method to return an object of type `KeysView` instance, and similary `DictConfig.items` now returns an `ItemsView`.
